transport/src/iosMain/kotlin/com/genesys/cloud/messenger/transport/network/PlatformSocket.kt

```diff
@@ -1,60 +1,123 @@
+import platform.Foundation.NSURLSessionWebSocketMessage
+import platform.Foundation.NSURLSessionWebSocketTask
+import platform.Foundation.setValue
+import platform.darwin.NSObject
+import platform.posix.ETIMEDOUT
 
 internal actual class PlatformSocket actual constructor(
     private val log: Log,
     configuration: Configuration,
+    /**
+     * Interval to automatically send pings while active. If pong not received within `interval`,
+     * client assumes connectivity is lost and will notify [PlatformSocketListener.onFailure].
+     */
     actual val pingInterval: Long
 ) {
     private val url = configuration.webSocketUrl
     private val socketEndpoint = NSURL.URLWithString(url.toString())!!
-    private var socket: JFRWebSocket? = null
+    private var webSocket: NSURLSessionWebSocketTask? = null
     private var pingTimer: NSTimer? = null
+    private var listener: PlatformSocketListener? = null
+    private val active: Boolean
+        get() = webSocket != null
 
     actual fun openSocket(listener: PlatformSocketListener) {
-        if (socket?.isConnected == true) {
-            listener.onFailure(Throwable("Socket is already connected."))
-            return
-        }
-
-        socket = JFRWebSocket(socketEndpoint, null)
-        socket?.onConnect = {
-            log.i { "onConnect()" }
-            listener.onOpen()
-        }
-        socket?.onDisconnect = { nsError ->
-            log.i { "onDisconnect(): $nsError" }
-            when {
-                nsError != null -> {
-                    listener.onFailure(Throwable(nsError.description))
+        val urlRequest = NSMutableURLRequest(socketEndpoint)
+        urlRequest.setValue(url.host, forHTTPHeaderField = "Origin")
+        val urlSession = NSURLSession.sessionWithConfiguration(
+            configuration = NSURLSessionConfiguration.defaultSessionConfiguration(),
+            delegate = object : NSObject(), NSURLSessionWebSocketDelegateProtocol {
+                override fun URLSession(
+                    session: NSURLSession,
+                    webSocketTask: NSURLSessionWebSocketTask,
+                    didOpenWithProtocol: String?
+                ) {
+                    log.i { "Socket did open." }
+                    listener.onOpen()
+                    keepAlive()
                 }
-            }
-            val closeCode =
-                if (nsError == null) SocketCloseCode.NORMAL_CLOSURE.value else SocketCloseCode.NO_STATUS_RECEIVED.value
-            listener.onClosed(closeCode, "disconnected")
-        }
-        socket?.onText = { text ->
-            text?.let { listener.onMessage(it) }
+                override fun URLSession(
+                    session: NSURLSession,
+                    webSocketTask: NSURLSessionWebSocketTask,
+                    didCloseWithCode: NSURLSessionWebSocketCloseCode,
+                    reason: NSData?
+                ) {
+                    val why = reason?.string() ?: ""
```

Comment:
Instead of empty string it might be more useful to say something like: "Reason not specified".